ZEP000-build-state-on-followers.md

```diff
@@ -0,0 +1,324 @@
+
+After switching to the Follower, we need to stop the processing, exporting and regularly sending of last exported positions. It might happen that we still write follow-up events to the log, because the RAFT role transition is already done, but the Zeebe Partition is notified asynchronous. This will be handled on the RAFT side, since it doesn't allow writing to the log if the node is not the leader. Still we should disallow new incoming user commands, via disabling the Command API. We could also think about having it always on, which would simplify it but this is out of scope.
+
+One important part before we start with the Follower "Replay Mode" is that we need to restore the current state from the latest snapshot. This is necessary because when we process commands we immediately apply the follow-up events to our state, which we write to the log later. We can't be sure what we have already applied is committed, and to not apply events twice it is easier to just restore from the latest snapshot, KISS principle. Other than that it is similar to described [above](#bootstrap-follower-partition).
+
+  
+[comment]: <> (          We have several challenges we need to overcome to solve/implement build state on followers. Let's take a look of the following causality chain:)
+[comment]: <> (           * To have a working and long living system, we need to be able to compact our log no matter the role. )
+[comment]: <> (           * To compact our log, we need to be able to take snapshots from the state.)
+[comment]: <> (             * The state need to contain a dense reflection of the log, last exported and processed position, from which we can recover.)
+[comment]: <> (           * To take a snasphot from the state, we need to build the state on all participants &#40;leader and follower&#41; in order to not replicate it.)
+[comment]: <> (           * To build state on followers, we need to avoid creating follow-up events by the state machine, Single Writer Principle. )
+[comment]: <> (           * To avoid follow-up events, we need to either replay only or use a NoopWriter.)
+[comment]: <> (In the following sections we will go in more depth in some these mentioned parts of the chain.)
+
+### Compacting the log
+
+In order to avoid an ever-growing log we need to compact it from time to time. This can be done by taking snapshots from the state, which allows us to start from that specific snapshot, such that we can compact our log. With building state on followers each node is responsible for taking their own snapshots and compacting their log.
+
+Part of the snapshot should be the last processed and last exported position for each exporter. The smallest number indicates until which position we can compact our log. In order to avoid running the exporters on all nodes and reduce the network load (since they export normally to an external system), the Leader has to periodically sync the last exported position to the followers. This is done via SWIM.
```

Comment:
What's wrong with producing an event on the log which can be applied to update the position? I see it mentioned in alternatives, but there's no explanation as to why it was discarded?

Reply:
I agree that we should discuss this point a bit more (in the ZEP).

Reply:
> Does it have to be SWIM, or should we send it via direct leader-follower communication via unicast or broadcast?

No doesn't need to be SWIM. I was also not sure to mention this here. It is one way to do it.

> What's wrong with producing an event on the log which can be applied to update the position? I see it mentioned in alternatives, but there's no explanation as to why it was discarded.

For me the sending of the state sound it simple enough for the first solution. It doesn't mean we can't improve. I can imagine that we could also iterate over it later and go with another solution. If you go directly with it it means we need to introduce a new record or value type (lot of protocol changes), have a writer in the exporter, and also a replay mechanism for the exporter, we need to ignore exporter records etc.

Reply:
> I agree that we should discuss this point a bit more (in the ZEP).

I would like to start first with the prototype and need to find out how it could really work. Maybe I see then some issues with the proposed solution, which I currently not seeing.

But in general I think this ZEP should focus on building state, which is highly focuses on the processing part and could be seen as the base. After that we can still improve and write a new ZEP which covers the exporter part more in depth. It is just a matter of being focused I would say. The exporter stuff has a rat tail. Otherwise I fear we spent to much time to figure it out all it once.